File: bindings/js-browser/src/tools/registry.js
/**
 * registry.js – ToolRegistry: maps tool names to definitions and handlers.
 *
 * Usage:
 *   const reg = new ToolRegistry();
 *   reg.register(
 *     { name: 'get_weather', description: '...', parameters: {} },
 *     async ({ city }) => ({ temperature: '30°C' })
 *   );
 *   const result = await reg.execute('get_weather', { city: 'Chennai' });
 *   const defs   = reg.definitions();
 */

export class ToolRegistry {
  /** @type {Map<string, {def: object, handler: Function}>} */
  #tools = new Map();
  /** @type {Array<object>} Preserves registration order for definitions(). */
  #order = [];

  /**
   * Register a tool.
   *
   * @param {object}   def          tool definition ({ name, description, parameters })
   * @param {Function} handler      async (args: object) => result
   * @throws {Error} if name is empty or handler is not a function
   */
  register(def, handler) {
    if (!def || !def.name) {
      throw new Error('ToolRegistry: tool name must not be empty');
    }
    if (typeof handler !== 'function') {
      throw new Error(`ToolRegistry: handler for "${def.name}" must be a function`);
    }

    if (!this.#tools.has(def.name)) {
      this.#order.push(def);
    } else {
      // Update ordered definition.
      const idx = this.#order.findIndex(d => d.name === def.name);
      if (idx !== -1) this.#order[idx] = def;
    }
    this.#tools.set(def.name, { def, handler });
  }

  /**
   * Execute a registered tool by name.
   *
   * @param {string} name  registered tool name
   * @param {object} args  decoded argument object from the model
   * @returns {Promise<*>} tool result (JSON-serialisable)
   * @throws {Error} if the tool is not registered
   */
  async execute(name, args) {
    const entry = this.#tools.get(name);
    if (!entry) {
      throw new Error(`ToolRegistry: unknown tool "${name}"`);
    }
    return entry.handler(args ?? {});
  }

  /**
   * Return all registered tool definitions in registration order.
   *
   * @returns {Array<object>}
   */
  definitions() {
    return this.#order.map(d => ({ ...d }));
  }

  /**
   * Check whether a tool with the given name is registered.
   *
   * @param {string} name
   * @returns {boolean}
   */
  has(name) {
    return this.#tools.has(name);
  }
}

File: bindings/go/tools/registry.go
// Package tools provides a ToolRegistry that maps tool names to their
// definitions and Go handler functions.
//
// The registry is safe for concurrent read access; writes (Register) should
// be done before the Agent starts handling requests.
package tools

import (
	"fmt"
	"sync"

	"github.com/muthuishere/llama-bindings/go/llama"
)

// Handler is a function that executes a tool call.
// args contains the decoded JSON arguments from the model.
// The return value is serialised back to JSON and forwarded to the model.
type Handler func(args map[string]interface{}) (interface{}, error)

// Registry maps tool names to their definitions and handlers.
type Registry struct {
	mu       sync.RWMutex
	tools    map[string]entry
	ordered  []llama.ToolDefinition // preserves registration order
}

type entry struct {
	def     llama.ToolDefinition
	handler Handler
}

// New creates an empty ToolRegistry.
func New() *Registry {
	return &Registry{tools: make(map[string]entry)}
}

// Register adds (or replaces) a tool.
// def.Name must be non-empty and unique; handler must not be nil.
func (r *Registry) Register(def llama.ToolDefinition, handler Handler) error {
	if def.Name == "" {
		return fmt.Errorf("tools: tool name must not be empty")
	}
	if handler == nil {
		return fmt.Errorf("tools: handler for %q must not be nil", def.Name)
	}

	r.mu.Lock()
	defer r.mu.Unlock()

	if _, exists := r.tools[def.Name]; !exists {
		r.ordered = append(r.ordered, def)
	} else {
		// Update existing entry in ordered slice.
		for i := range r.ordered {
			if r.ordered[i].Name == def.Name {
				r.ordered[i] = def
				break
			}
		}
	}
	r.tools[def.Name] = entry{def: def, handler: handler}
	return nil
}

// Definitions returns all registered tool definitions in registration order.
func (r *Registry) Definitions() []llama.ToolDefinition {
	r.mu.RLock()
	defer r.mu.RUnlock()
	out := make([]llama.ToolDefinition, len(r.ordered))
	copy(out, r.ordered)
	return out
}

// Execute calls the handler registered for name with the provided args.
// Returns an error if the tool is not found or the handler returns one.
func (r *Registry) Execute(name string, args map[string]interface{}) (interface{}, error) {
	r.mu.RLock()
	e, ok := r.tools[name]
	r.mu.RUnlock()
	if !ok {
		return nil, fmt.Errorf("tools: unknown tool %q", name)
	}
	return e.handler(args)
}

// Has reports whether a tool with the given name is registered.
func (r *Registry) Has(name string) bool {
	r.mu.RLock()
	_, ok := r.tools[name]
	r.mu.RUnlock()
	return ok
}
